### PR TITLE
feat(*): add leader election example

### PR DESCRIPTION
**Pull request by bbbmj**
**What this PR does / why we need it**:

add leader election example

先前的 PR：https://github.com/caicloud/go-common/pull/82

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @supereagle @lsytj0413 @ddysher 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,26 +2,52 @@
...
 
-## About the project
+This example demonstrates how to use the leader election package.
 
-The Common Template Project is a template for starting new projects at Caicloud. All Caicloud projects, at minimum, must have the following files:
+## Running
 
-- a `README.md` outlining the project
-- an `OWNERS` with the project leads listed as approvers ([docs on `OWNERS` files](https://github.com/caicloud/engineering/blob/master/guidelines/caicloud_bot.md#the-owners-file))
+Run the following three commands in separate terminals.
```
Comment (supereagle) [inline README.md:17]:
```suggestion
Run the following three commands in separate terminals of one machine. If you run them on different machines, you can use the same port.
```

File: main.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	var kubeconfig string
+	var leaseLockName string
+	var leaseLockNamespace string
+	var id string
+
+	flag.IntVar(&port, "port", 8080, "port to listen")
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
+	flag.StringVar(&id, "id", "", "the holder identity name")
+	flag.StringVar(&leaseLockName, "lease-lock-name", "", "the lease lock resource name")
+	flag.StringVar(&leaseLockNamespace, "lease-lock-namespace", "", "the lease lock resource namespace")
```
Comment (supereagle) [inline main.go:68]: Give default values for these two flags?

File: main.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		select {}
+	}
+
+	klog.Infof("Run with leader election")
+
+	// we use the Lease lock type since edits to Leases are less common
+	// and fewer objects in the cluster watch "all Leases".
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      leaseLockName,
```
Comment (supereagle) [inline main.go:144]: Whether the lease lock needs to be created in advance?

File: main.go (hunk 2 above)
Comment (bbbmj) [reply main.go:68]: > I would drop the default lease lock name and namespace and require values be passed in via args. Assuming people will copy this verbatim, we don't want lots of random controllers competing on the example/default name/namespace

File: main.go (hunk 3 above)
Comment (bbbmj) [reply main.go:144]: no

**Comment by deyuanix** [comment]
@iawia002 

/cc @lsytj0413 @zxq-bit

**Comment by lsytj0413** [comment]
LGTM

**Pull request merged**